sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala

```diff
@@ -47,4 +48,44 @@ case class Md5(child: Expression)
       "org.apache.spark.unsafe.types.UTF8String.fromString" +
         s"(org.apache.commons.codec.digest.DigestUtils.md5Hex($c))")
   }
+
+}
+
+/**
+ * A function that returns a hash value of the argument
+ */
+case class Hash(children: Expression*) extends Expression {
+
+  override def foldable: Boolean = children.forall(_.foldable)
+  override def nullable: Boolean = children(0).nullable
+  override def dataType: DataType = IntegerType
+
+  override def eval(input: InternalRow): Any = {
+    val arglist = children.map(_.eval(input).asInstanceOf[AnyRef])
```

Comment:
Could we make argList as InternalRow, then we can use InternalRow.hashCode()?

Thinking of let `hash(A, B, C)` be consistent as `group by A, B, C`, we can know the partition after shuffle by `hash(A, B, C) % numPartitions`